### PR TITLE
#74 toast component

### LINKED ISSUE #74: Alert component
### Alert component

There is a new design specification for the alert component, that is used across the application. The alert component is used to display messages and give feedback of actions to the user, such as success messages, error messages, and warnings.

#### Design reference
The design specifications for the new design can be found on the following Figma link:
https://www.figma.com/design/XqVOGOoS8tZhGK5ROmqGFv/Open-tasks?node-id=131-2192&node-type=frame&t=ylJesJp7IAC7tzGk-0

#### Development
Under the hood, the alert component is based on the Toast component provided by shadcn. Use can use the base functionality and only update the styles to match our design spec.

The existing component documentation can be found on the following link:
https://ui.shadcn.com/docs/components/toast

There is no need for much flexibility in this component, as it is used for specific purposes throughout the application, so only the existing properties are really necessary and must be kept as is.

Also notice that there is no size variants, so the set of variants can be reduced down to: `info`, `success`, `warning`, and `error`.

It's important to have flexibility in the positioning of the alert component, so it can be used in different parts of the application. The component should be able to be positioned at the top, bottom, left, and right of the screen.

Also notice that the close button functionality is already implemented, so it's only a matter of customizing to match the specifications.

#### Acceptance criteria
- [ ] The component variants must match the ones in the design specs: `info`, `success`, `warning`, and `error`.
- [ ] The component must be able to be positioned at the top, bottom, left, and right of the screen.
- [ ] The component must be documented with stories in Storybook.

## Important: branch from `main`

### PR DESCRIPTION
**Pull request by nafisreza**
### Changes:
- Supports 4 different variants (info, success, warning, and error)
- Can be positioned on different places of the screen (top-left, bottom-left, top-right, bottom-right)
- Size, colors, icons according to the Figma design. 

### How to use:
``` 
<Button
      onClick={() => {
        toast({
          title: "Alert message goes here!",
          description: "This is where a alert description goes, replace it here",
          variant: "success",
          position: "bottom-right",
          duration: 2000,
        })
      }}
    >
      Show Toast
</Button>
```

### Screenshots:

https://github.com/user-attachments/assets/8172cbbb-f96a-4072-871c-c8119d43f407

### CODE REVIEW

**Comment by nafisreza** [comment]
One thing that can be further improved here is when the toast is positioned on the left, we can animate the closing in the same direction or it may be good as it is (whatever looks cool). Currently, it is using the translation x/y from radix-toast-swipe. Then the animation would need to be written manually.

File: src/components/ui/toast.tsx [hunk 1]
```diff
@@ -9,31 +9,50 @@ const ToastProvider = ToastPrimitives.Provider
...
+      case 'top-right':
+        return 'top-0 right-0';
+      case 'bottom-left':
+        return 'bottom-0 left-0';
+      case 'bottom-right':
+        return 'bottom-0 right-0';
+      default:
+        return 'top-0 right-0';
+    }
+  })();
```
Comment (cryptosalomao) [inline src/components/ui/toast.tsx:28]: Here is a cleaner approach:

```typescript
const positionStyles = {
    'top-left': 'top-0 left-0',
    'top-right': 'top-0 right-0',
    'bottom-left': 'bottom-0 left-0',
    'bottom-right': 'bottom-0 right-0',
}[position] || 'top-0 right-0';
```

File: src/components/ui/toast.tsx [hunk 2]
```diff
@@ -9,31 +9,50 @@ const ToastProvider = ToastPrimitives.Provider
...
+
+  return (
+    <ToastPrimitives.Viewport
+      ref={ref}
+      className={cn(
+        `fixed z-[100] flex max-h-screen w-full flex-col-reverse p-4 sm:flex-col md:max-w-[420px] ${positionStyles}`,
```
Comment (cryptosalomao) [inline src/components/ui/toast.tsx:34]: `z-10` is sufficient here

File: src/components/ui/toast.tsx [hunk 3]
```diff
@@ -9,31 +9,50 @@ const ToastProvider = ToastPrimitives.Provider
...
-          "destructive group border-destructive bg-destructive text-destructive-foreground",
+        success: "bg-elevation-200 text-foreground",
+        info: "bg-elevation-200 text-foreground",
+        warning: "bg-elevation-200 text-foreground",
+        error: "bg-elevation-200 text-foreground",
       },
     },
     defaultVariants: {
-      variant: "default",
+      variant: "success",
```
Comment (cryptosalomao) [inline src/components/ui/toast.tsx:55]: `info` as a default variant

File: src/components/ui/toast.tsx [hunk 4]
```diff
@@ -81,7 +100,7 @@ const ToastClose = React.forwardRef<
     toast-close=""
     {...props}
   >
-    <X className="h-4 w-4" />
+    <X className="h-5 w-5 text-text-200" />
```
Comment (cryptosalomao) [inline src/components/ui/toast.tsx:103]: missing `strokeWidth` prop here, must be 1px thick

File: src/components/ui/toast.tsx (hunk 2 above)
Comment (nafisreza) [reply src/components/ui/toast.tsx:34]: Right. Actually this value is from shadcn itself.

File: src/components/ui/toast.tsx [hunk 5]
```diff
@@ -92,7 +111,7 @@ const ToastTitle = React.forwardRef<
 >(({ className, ...props }, ref) => (
   <ToastPrimitives.Title
     ref={ref}
-    className={cn("text-sm font-semibold", className)}
+    className={cn("text-[14px] text-text-200 font-[500]", className)}
```
Comment (cryptosalomao) [inline src/components/ui/toast.tsx:114]: `text-sm` and `font-medium` in this line

File: src/components/ui/toaster.tsx [hunk 6]
```diff
@@ -1,46 +1,52 @@
...
+      case 'info':
+        return <InfoIcon className="h-5 w-5" />
+      case 'warning':
+        return <TriangleAlertIcon className="h-5 w-5 text-signal-alert" />
+      case 'error':
+        return <OctagonXIcon className="h-5 w-5 text-signal-error" />
+      default:
+        return null
+    }
+  }
```
Comment (cryptosalomao) [inline src/components/ui/toaster.tsx:28]: Cleaner version:

```typescript
const icons = {
    success: <CircleCheckIcon className="w-4 h-4 text-signal-success" />,
    info: <InfoIcon className="w-4 h-4" />,
    warning: <TriangleAlertIcon className="w-4 h-4 text-signal-alert" />,
    error: <OctagonXIcon className="w-4 h-4 text-signal-error" />,
};

const icon = icons[variant] || icons.info;
```

And notice the `strokeWidth` and the icon dimensions too

File: src/components/ui/toaster.tsx [hunk 7]
```diff
@@ -1,46 +1,52 @@
...
   return (
     <ToastProvider>
-      {toasts.map(function ({
-        id,
-        title,
-        description,
-        action,
-        ...props
-      }: ToasterProps) {
+      {toasts.map(function ({ id, title, description, action, variant, position, ...props }) {
```
Comment (cryptosalomao) [inline src/components/ui/toaster.tsx:32]: `position` is unused here

File: src/hooks/use-toast.ts [hunk 8]
```diff
@@ -1,101 +1,104 @@
 "use client";
```
Comment (cryptosalomao) [inline src/hooks/use-toast.ts:1]: This line can go away

File: src/hooks/use-toast.ts [hunk 9]
```diff
@@ -1,101 +1,104 @@
...
+  action?: ToastActionElement
+  position?: string
+}
+
+const actionTypes = {
+  ADD_TOAST: "ADD_TOAST",
+  UPDATE_TOAST: "UPDATE_TOAST",
+  DISMISS_TOAST: "DISMISS_TOAST",
+  REMOVE_TOAST: "REMOVE_TOAST",
+} as const
```
Comment (cryptosalomao) [inline src/hooks/use-toast.ts:26]: This has been already fixed. Please rebase main into your branch

File: src/hooks/use-toast.ts [hunk 10]
```diff
@@ -1,101 +1,104 @@
...
+      type: ActionType["DISMISS_TOAST"]
+      toastId?: ToasterToast["id"]
     }
   | {
-      type: ActionType.REMOVE_TOAST;
-      toastId?: ToasterToast["id"];
-    };
+      type: ActionType["REMOVE_TOAST"]
+      toastId?: ToasterToast["id"]
+    }
```
Comment (cryptosalomao) [inline src/hooks/use-toast.ts:53]: keep the enum

**Review by cryptosalomao — CHANGES_REQUESTED** [review]
Hey @nafisreza very good one! Thanks for the contrib. I've left some important points to be addressed.

Please notice that a more recent version has been merged to `main` in order to fix some type warnings, so rebase `main` into your branch. And it's important to stick with the project's conventions for eslint, so please configure your IDE to not override them. Lots of semicolons have been removed unnecessarily.

**Comment by mtbitcr** [comment]
@nafisreza we will announce PRs Rules soon. About PR naming, necessary fields to fill in and so on

**Comment by mtbitcr** [comment]
@nafisreza I don't understand to what task this PR belong. Please stick to naming #TASKNUMBER_TASKNAME. And also attach task in Development.

**Comment by mtbitcr** [comment]
Hey @nafisreza . Please re-request review after fixing comments.

**Review by cryptosalomao — APPROVED** [review]
LGTM

**Pull request merged**